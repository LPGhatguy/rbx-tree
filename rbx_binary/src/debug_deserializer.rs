//! Deserializer that reads a file and creates a debug representation of it.
//! It's intended to be used to snapshot test the binary serializer without
//! suffering from same-inverse-bug problems.

use std::io::Read;

use byteorder::{LittleEndian, ReadBytesExt};
use serde::{Deserialize, Serialize};

use crate::{chunk::Chunk, core::RbxReadExt, deserializer::FileHeader};

#[derive(Debug, Serialize, Deserialize)]
pub struct DecodedModel {
    pub num_types: u32,
    pub num_instances: u32,
    pub chunks: Vec<DecodedChunk>,
}

impl DecodedModel {
    pub fn from_reader<R: Read>(mut reader: R) -> Self {
        let header = FileHeader::decode(&mut reader).expect("invalid file header");
        let mut chunks = Vec::new();

        loop {
            let chunk = Chunk::decode(&mut reader).expect("invalid chunk");

            match &chunk.name {
                b"META" => chunks.push(decode_meta_chunk(chunk.data.as_slice())),
                b"INST" => chunks.push(decode_inst_chunk(chunk.data.as_slice())),
                b"PROP" => chunks.push(decode_prop_chunk(chunk.data.as_slice())),
                b"PRNT" => chunks.push(decode_prnt_chunk(chunk.data.as_slice())),
                b"END\0" => {
                    chunks.push(DecodedChunk::End);
                    break;
                }
                _ => {
                    chunks.push(DecodedChunk::Unknown {
                        name: String::from_utf8_lossy(&chunk.name[..]).to_string(),
                        contents: chunk.data,
                    });
                }
            }
        }

        DecodedModel {
            num_types: header.num_types,
            num_instances: header.num_instances,
            chunks,
        }
    }
}

fn decode_meta_chunk<R: Read>(mut reader: R) -> DecodedChunk {
    let num_entries = reader.read_u32::<LittleEndian>().unwrap();
    let mut entries = Vec::with_capacity(num_entries as usize);

    for _ in 0..num_entries {
        let key = reader.read_string().unwrap();
        let value = reader.read_string().unwrap();
        entries.push((key, value));
    }

    let mut remaining = Vec::new();
    reader.read_to_end(&mut remaining).unwrap();

    DecodedChunk::Meta { entries, remaining }
}

fn decode_inst_chunk<R: Read>(mut reader: R) -> DecodedChunk {
    let type_id = reader.read_u32::<LittleEndian>().unwrap();
    let type_name = reader.read_string().unwrap();
    let object_format = reader.read_u8().unwrap();
    let num_instances = reader.read_u32::<LittleEndian>().unwrap();

    let mut referents = vec![0; num_instances as usize];
    reader.read_referent_array(&mut referents).unwrap();

    let mut remaining = Vec::new();
    reader.read_to_end(&mut remaining).unwrap();

    DecodedChunk::Inst {
        type_id,
        type_name,
        object_format,
        referents,
        remaining,
    }
}

fn decode_prop_chunk<R: Read>(mut reader: R) -> DecodedChunk {
    let type_id = reader.read_u32::<LittleEndian>().unwrap();
    let prop_name = reader.read_string().unwrap();
    let prop_type = reader.read_u8().unwrap();

    let mut remaining = Vec::new();
    reader.read_to_end(&mut remaining).unwrap();

    DecodedChunk::Prop {
        type_id,
        prop_name,
        prop_type,
        remaining,
    }
}

fn decode_prnt_chunk<R: Read>(mut reader: R) -> DecodedChunk {
    let version = reader.read_u8().unwrap();
    let num_referents = reader.read_u32::<LittleEndian>().unwrap();

    let mut subjects = vec![0; num_referents as usize];
    let mut parents = vec![0; num_referents as usize];

    reader.read_referent_array(&mut subjects).unwrap();
    reader.read_referent_array(&mut parents).unwrap();

    let links = subjects
        .iter()
        .copied()
        .zip(parents.iter().copied())
        .collect();

    let mut remaining = Vec::new();
    reader.read_to_end(&mut remaining).unwrap();

    DecodedChunk::Prnt {
        version,
        links,
        remaining,
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub enum DecodedChunk {
    Meta {
        entries: Vec<(String, String)>,
        #[serde(with = "unknown_buffer")]
        remaining: Vec<u8>,
    },

    Inst {
        type_id: u32,
        type_name: String,
        object_format: u8,
        referents: Vec<i32>,
        #[serde(with = "unknown_buffer")]
        remaining: Vec<u8>,
    },

    Prop {
        type_id: u32,
        prop_name: String,
        prop_type: u8,
        #[serde(with = "unknown_buffer")]
        remaining: Vec<u8>,
    },

    Prnt {
        version: u8,
        links: Vec<(i32, i32)>,
        #[serde(with = "unknown_buffer")]
        remaining: Vec<u8>,
    },

    End,

    Unknown {
        name: String,
        #[serde(with = "unknown_buffer")]
        contents: Vec<u8>,
    },
}

/// Contains data that we haven't decoded for a chunk. Using `unknown_buffer`
/// should generally be a placeholder since it's results are opaque, but stable.
mod unknown_buffer {
    use serde::{Deserialize, Deserializer, Serializer};

    pub fn serialize<S>(value: &Vec<u8>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.collect_str(&base64::display::Base64Display::with_config(
            &value,
            base64::STANDARD,
        ))
    }

    pub fn deserialize<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let encoded = <&str>::deserialize(deserializer)?;
        let contents = base64::decode(encoded).map_err(serde::de::Error::custom)?;

        Ok(contents)
    }
}
