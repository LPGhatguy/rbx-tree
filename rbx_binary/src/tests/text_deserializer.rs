use crate::debug_deserializer::DecodedModel;

#[test]
fn folder_from_studio() {
    static FOLDER: &[u8] = include_bytes!("../../test-files/just-folder.rbxm");

    let decoded = DecodedModel::from_reader(FOLDER);
    insta::assert_yaml_snapshot!(decoded);
}
