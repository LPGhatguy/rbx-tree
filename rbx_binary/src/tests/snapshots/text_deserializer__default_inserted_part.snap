---
source: rbx_binary/src/tests/text_deserializer.rs
expression: decoded
---
num_types: 1
num_instances: 1
chunks:
  - Meta:
      entries:
        - - ExplicitAutoJoints
          - "true"
  - Inst:
      type_id: 0
      type_name: Part
      object_format: 0
      referents:
        - 0
  - Prop:
      type_id: 0
      prop_name: Anchored
      prop_type: 2
      remaining: AA==
  - Prop:
      type_id: 0
      prop_name: AttributesSerialize
      prop_type: 1
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: BackParamA
      prop_type: 4
      remaining: fgAAAQ==
  - Prop:
      type_id: 0
      prop_name: BackParamB
      prop_type: 4
      remaining: fgAAAA==
  - Prop:
      type_id: 0
      prop_name: BackSurface
      prop_type: 18
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: BackSurfaceInput
      prop_type: 18
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: BottomParamA
      prop_type: 4
      remaining: fgAAAQ==
  - Prop:
      type_id: 0
      prop_name: BottomParamB
      prop_type: 4
      remaining: fgAAAA==
  - Prop:
      type_id: 0
      prop_name: BottomSurface
      prop_type: 18
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: BottomSurfaceInput
      prop_type: 18
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: CFrame
      prop_type: 16
      remaining: AoGAAAF+AAAggoAAAQ==
  - Prop:
      type_id: 0
      prop_name: CanCollide
      prop_type: 2
      remaining: AQ==
  - Prop:
      type_id: 0
      prop_name: CastShadow
      prop_type: 2
      remaining: AQ==
  - Prop:
      type_id: 0
      prop_name: CollisionGroupId
      prop_type: 3
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: Color3uint8
      prop_type: 26
      remaining: o6Kl
  - Prop:
      type_id: 0
      prop_name: CustomPhysicalProperties
      prop_type: 25
      remaining: AA==
  - Prop:
      type_id: 0
      prop_name: FrontParamA
      prop_type: 4
      remaining: fgAAAQ==
  - Prop:
      type_id: 0
      prop_name: FrontParamB
      prop_type: 4
      remaining: fgAAAA==
  - Prop:
      type_id: 0
      prop_name: FrontSurface
      prop_type: 18
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: FrontSurfaceInput
      prop_type: 18
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: LeftParamA
      prop_type: 4
      remaining: fgAAAQ==
  - Prop:
      type_id: 0
      prop_name: LeftParamB
      prop_type: 4
      remaining: fgAAAA==
  - Prop:
      type_id: 0
      prop_name: LeftSurface
      prop_type: 18
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: LeftSurfaceInput
      prop_type: 18
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: Locked
      prop_type: 2
      remaining: AA==
  - Prop:
      type_id: 0
      prop_name: Massless
      prop_type: 2
      remaining: AA==
  - Prop:
      type_id: 0
      prop_name: Material
      prop_type: 18
      remaining: AAABAA==
  - Prop:
      type_id: 0
      prop_name: Name
      prop_type: 1
      remaining: BAAAAFBhcnQ=
  - Prop:
      type_id: 0
      prop_name: Reflectance
      prop_type: 4
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: RightParamA
      prop_type: 4
      remaining: fgAAAQ==
  - Prop:
      type_id: 0
      prop_name: RightParamB
      prop_type: 4
      remaining: fgAAAA==
  - Prop:
      type_id: 0
      prop_name: RightSurface
      prop_type: 18
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: RightSurfaceInput
      prop_type: 18
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: RootPriority
      prop_type: 3
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: RotVelocity
      prop_type: 14
      remaining: AAAAAAAAAAAAAAAA
  - Prop:
      type_id: 0
      prop_name: Tags
      prop_type: 1
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: TopParamA
      prop_type: 4
      remaining: fgAAAQ==
  - Prop:
      type_id: 0
      prop_name: TopParamB
      prop_type: 4
      remaining: fgAAAA==
  - Prop:
      type_id: 0
      prop_name: TopSurface
      prop_type: 18
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: TopSurfaceInput
      prop_type: 18
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: Transparency
      prop_type: 4
      remaining: AAAAAA==
  - Prop:
      type_id: 0
      prop_name: Velocity
      prop_type: 14
      remaining: AAAAAAAAAAAAAAAA
  - Prop:
      type_id: 0
      prop_name: formFactorRaw
      prop_type: 18
      remaining: AAAAAQ==
  - Prop:
      type_id: 0
      prop_name: shape
      prop_type: 18
      remaining: AAAAAQ==
  - Prop:
      type_id: 0
      prop_name: size
      prop_type: 14
      remaining: gQAAAH8AAACAAAAA
  - Prnt:
      version: 0
      links:
        - - 0
          - -1
  - End
